### PR TITLE
A cleaner implementation of the tab complete option

### PR DESCRIPTION
**Pull request by Konfekt**
At same point, replace `inoremap <tab> ...` by `imap <tab> ..` to allow for say `z` used by https://github.com/chrisbra/unicode.vim

### CODE REVIEW

**Comment by Konfekt** [comment]
I am sorry, at some point this became a major refactoring. But all good ! Check it out if it suits you or if it should live as a separate fork.

**Comment by ajh17** [comment]
Hey, I welcome all refactoring efforts, thanks! :smile: 

I will take some days to go through the change and test it out myself though, but looking at it briefly, it looks pretty good.

**Comment by Konfekt** [comment]
Hey, I got carried away with this and made a complete rewrite that, however, currently mimics VCM's defaults. Let me present it to you in another pull request. 

I will close this one, like the former, but both are closer to VimCompletesMe and therefore perhaps more to your liking. For the mean while, I'll close these.

**Comment by ajh17** [comment]
@Konfekt I actually like this pull request more, and would prefer #24 to live as a separate fork.

File: plugin/VimCompletesMe.vim [hunk 1]
```diff
@@ -23,64 +23,57 @@ if !exists('g:vcm_default_maps')
 endif
 
 if !exists('g:vcm_omni_pattern')
-  let g:vcm_omni_pattern = '\v(\.\|->\|::)'
+  let g:vcm_omni_pattern = '\v\k+(\.|->|::)\k*$'
```
Comment (ajh17) [inline plugin/VimCompletesMe.vim]: I'm just noticing now but can you remove the `\v` usages? I don't like very magic mode myself, and I believe Vim plugins should use the default `\m` mode.

File: doc/VimCompletesMe.txt [hunk 2]
```diff
@@ -34,41 +35,41 @@ following options:
...
+
+In some situations (like after completing something and then trying the
+context switch again without leaving insert mode), you may need a Vim
+version greater than 7.3.598 to get the best context switch behavior.
+
+                                                         *'g:vcm_tab_complete'*
+Values: string                                                                ~
+Default: ''                                                                   ~
+
+The global variant of |b:vcm_tab_complete| and which which applies in case
```
Comment (ajh17) [inline doc/VimCompletesMe.txt]: "which which" is repeated here.

File: doc/VimCompletesMe.txt [hunk 3]
```diff
@@ -110,18 +111,22 @@ own mappings, unset this variable by putting the following in your |vimrc|:
...
+The buffer-local pattern after which VimCompletesMe triggers Omni-completion.
+To be set by an 'autocmd FileType ...'. Set it equal to '0' to make it never
+match.
 
 ------------------------------------------------------------------------------
-                                                         *'b:vcm_omni_pattern*
+                                                         *'g:vcm_omni_pattern*
 Values: string                                                               ~
-Default: g:vcm_omni_pattern                                                  ~
+Default: '\v\k+(\.|->|::)\k*$'                                                ~
```
Comment (ajh17) [inline doc/VimCompletesMe.txt]: Please avoid using `\v` mode.

File: plugin/VimCompletesMe.vim [hunk 4]
```diff
@@ -23,64 +23,57 @@ if !exists('g:vcm_default_maps')
...
-    endif
+      return Ctrl_PN
   endif
 
-  " Figure out whether we should indent.
+  " Cursor after whitespace => indent
   let pos = getpos('.')
   let substr = matchstr(strpart(getline(pos[1]), 0, pos[2]-1), "[^ \t]*$")
-  if strlen(substr) == 0
+  if empty(substr)
```
Comment (ajh17) [inline plugin/VimCompletesMe.vim]: Nice. Didn't know about `empty()`, thanks!

File: plugin/VimCompletesMe.vim [hunk 5]
```diff
@@ -23,64 +23,57 @@ if !exists('g:vcm_default_maps')
...
+  " Start completing => keep record for fallback
+  let b:tab_complete_pos = pos
   let b:completion_tried = 1
-  if map ==? "user"
-    return "\<C-x>\<C-u>"
-  elseif map ==? "omni"
+
+  " User is typing a path or omnicomplete or tabcomplete pattern?
+  let omni_pattern = get(b:, 'vcm_omni_pattern', get(g:, 'vcm_omni_pattern'))
+  let is_omni_pattern = (omni_pattern isnot 0) && (match(substr, omni_pattern) >= 0)
```
Comment (ajh17) [inline plugin/VimCompletesMe.vim]: Am I missing something? Why isnot used here instead of !=. Why are we comparing instance here?

**Comment by ajh17** [comment]
A couple of comments, but the change overall looks good to me. 

Also please avoid using very magic mode. It is recommended to always use the default `\m` mode. (`:help \v` --> read the note at the end)

File: plugin/VimCompletesMe.vim (hunk 5 above)
Comment (Konfekt) [inline plugin/VimCompletesMe.vim]: Ahh... I found it out the hard way. Try it, `!= 0` returns true if you compare the string on the left-hand side with the number 0 on the right-hand side.

**Comment by Konfekt** [comment]
Ok, https://github.com/ajh17/VimCompletesMe/pull/22 incorporates your suggestions.

**Pull request closed without merging**